FITB: Refactoring dynamic-feedback and defensive JSON creation

dbrianwalton opened the pull request:
Refactored `mode='dynamic-feedback'` from just matching `#fillin` to reference `#evaluate` so that the feedback evaluation is directly accessed instead of using exsl:node-tree. This allows localization within feedback creation since now the template has access to the language attribute of the original node-tree.

Removed `mode='body'` in template call on feedback element (was missed on an earlier clean-up)

Went through all of the `<xsl:choose>` and ensure that either `<xsl:otherwise>` was present, or not necessary, or flagged an error. This is the underlying cause of #2458. A problem that lacks a mode will now trigger an error, as will using undefined modes or missing or mismatching `#evaluate` elements.

rbeezer commented:
Looks good, but I think there is one oversight someplace.  Sample book, HTML output, diff before/after PR. Seems there are two "incorrect" strings, rather than a "correct" and an "incorrect".  Likely I could find it, but I bet you'll go right there.  Please add a **new** commit here, and I'll get in merged into the right place.  Long version for content, then short version.

```json
"problemHtml": "<div class=\"para\" id=\"ex-demo-jscmp-primes-2-1\">What is an example of a prime number less than 20? <input type=\"text
\" id=\"ex-demo-jscmp-primes-blank1\" size=\"5\"\/><div class=\"autopermalink\" data-description=\"Paragraph\"><a href=\"#ex-demo-jscmp-p
rimes-2-1\" title=\"Copy heading and permalink for Paragraph\">🔗<\/a><\/div><\/div>",
"solutionHtml": "",
"blankNames": {"blank1": 0},
"feedbackArray": [[{"solution_code": "function() {\n    var testResults = new Array();\n    testResults[0] = \n                        [2
, 3, 5, 7, 11, 13, 17, 19].includes(Number(ans))\n                        ;\n    return (testResults[0]);\n}()", "feedback": "<div class=
\"para\" id=\"ex-demo-jscmp-primes-3-1-1-2-1\">Any number from the list <span class=\"process-math\">\\(\\{2, 3, 5, 7, 11, 13, 17, 19\\}\
\)<\/span> is a prime number less than 20.<\/div>"}, {"solution_code": "function() {\n    var testResults = new Array();\n    testResults
[0] = \n                        function(){\n                        const val=Number(ans);\n                        if (val <= 0) { retu
rn \"You need to give a positive integer.\" }\n                        if (val >= 20) { return \"The integer must be less than 20.\" }\n 
                       for (let i=2; i<=3; i++) {\n                            if (val % i == 0) { return `Your answer is composite; for 
example, it is divisible by ${i}.` }\n                        }\n                        return false;\n                        }()\n    
                    ;\n    return (testResults[0]);\n}()", "feedback": [-"Correct!"},-]{+"Incorrect."},+} {"feedback": "Incorrect."}]]
}</script></div>
</div></div>
```

See near the end, with diff markers `([-,-]; {+,+}'
```json
"feedback": [-"Correct!"},-]{+"Incorrect."},+} {"feedback": "Incorrect."}
```

rbeezer commented:
Two comments, while I'm here.  These can wait, do not address them here - I'm ready but for change above.

* Those are very long commit messages!  I will edit.
* Nearby to some changes I saw several:  `xsl:text`, `xsl:value-of`, `xsl:text` all on one line.  I think the code is more readable if they were split onto three lines.  You won't get intervening whitespace into your output, it is ignored.

rbeezer commented:
One more thing!  Sorry!

* There is a `mode="location-report"` template which you can use after error messages.  Would save you reporting `visible-id`, and may generally be more helpful to authors.  Again, address that later on another PR.

dbrianwalton commented:
The new version is actually correct here. There was an error that I had fixed but did not realize was present in any of the examples. The `#test` that is being evaluated at that stage for feedback is actually the 2nd test and the javascript function is determining which specific feedback should be given for incorrect responses. If the function is unable to select a specific response, then the `feedback` should be used as generic feedback and should have been `Incorrect`. 

The error was that there had been a test on the variable `$b-correct` that only checked `test="$b-correct"` instead of looking at the actual value `yes` or `no`. It was fixed in commit 6410c7eb when I switched to using localized response strings.

rbeezer commented:
> The new version is actually correct here. 

Aah!  Good I didn't go looking to fix it myself.  ;-)  Will merge just now.
